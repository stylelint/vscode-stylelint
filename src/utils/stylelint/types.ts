import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line n/no-unpublished-import
import type stylelint from 'stylelint';
import type { PackageManager } from '../packages/index';
export type Stylelint = typeof stylelint;
export type ConfigurationError = Error & { code: 78 };

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;

	/**
	 * Gets the original warning from the given diagnostic.
	 */
	getWarning?: (diagnostic: LSP.Diagnostic) => stylelint.Warning | null;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}
