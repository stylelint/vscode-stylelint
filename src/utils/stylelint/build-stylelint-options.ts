import path from 'path';
import pathIsInside from 'path-is-inside';
import { URI } from 'vscode-uri';
import { findPackageRoot } from '../packages/index';
import type stylelint from 'stylelint';
import type { RunnerOptions } from './types';

/**
 * Given a document URI, base options, and extension options, builds a Stylelint
 * options object. Runner options supersede base options.
 */
export async function buildStylelintOptions(
	uri: string,
	workspaceFolder?: string,
	baseOptions: Partial<stylelint.LinterOptions> = {},
	{
		config,
		configFile,
		configBasedir,
		customSyntax,
		ignoreDisables,
		reportDescriptionlessDisables,
		reportNeedlessDisables,
		reportInvalidScopeDisables,
	}: RunnerOptions = {},
): Promise<Partial<stylelint.LinterOptions>> {
	const options = {
		...baseOptions,

		config: config ?? baseOptions.config,

		configFile: configFile
			? workspaceFolder
				? configFile.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: configFile
			: baseOptions.configFile,

		configBasedir: configBasedir
			? path.isAbsolute(configBasedir)
				? configBasedir
				: path.join(workspaceFolder ?? '', configBasedir)
			: baseOptions.configBasedir,

		customSyntax: customSyntax
			? workspaceFolder
				? customSyntax.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: customSyntax
			: baseOptions.customSyntax,

		ignoreDisables: ignoreDisables ?? baseOptions.ignoreDisables,

		reportDescriptionlessDisables:
			reportDescriptionlessDisables ?? baseOptions.reportDescriptionlessDisables,

		reportNeedlessDisables: reportNeedlessDisables ?? baseOptions.reportNeedlessDisables,

		reportInvalidScopeDisables:
			reportInvalidScopeDisables ?? baseOptions.reportInvalidScopeDisables,
	};

	const documentPath = URI.parse(uri).fsPath;

	if (documentPath) {
		if (workspaceFolder && pathIsInside(documentPath, workspaceFolder)) {
			options.ignorePath = path.join(workspaceFolder, '.stylelintignore');
		}

		if (options.ignorePath === undefined) {
			options.ignorePath = path.join(
				(await findPackageRoot(documentPath)) || path.parse(documentPath).root,
				'.stylelintignore',
			);
		}
	}

	return options;
}
