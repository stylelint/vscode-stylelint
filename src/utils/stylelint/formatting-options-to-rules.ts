import type LSP from 'vscode-languageserver-protocol';

export type FormattingRules = {
	indentation?: [number | string];
	'no-missing-end-of-source-newline'?: true | null;
	'no-eol-whitespace'?: true | null;
};

/**
 * Converts the given formatting options to rules.
 * @param options The formatting options.
 * @returns The rules.
 */
export function formattingOptionsToRules({
	insertSpaces,
	tabSize,
	insertFinalNewline,
	trimTrailingWhitespace,
}: LSP.FormattingOptions): FormattingRules {
	// NOTE: There is no equivalent rule for trimFinalNewlines, so it is not respected.
	// TODO: Create respective rule upstream?

	const rules: FormattingRules = {
		indentation: [insertSpaces ? tabSize : 'tab'],
	};

	if (insertFinalNewline !== undefined) {
		rules['no-missing-end-of-source-newline'] = insertFinalNewline || null;
	}

	if (trimTrailingWhitespace !== undefined) {
		rules['no-eol-whitespace'] = trimTrailingWhitespace || null;
	}

	return rules;
}
