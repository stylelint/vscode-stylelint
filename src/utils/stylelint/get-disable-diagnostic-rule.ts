import type LSP from 'vscode-languageserver-protocol';
import { DisableReportRuleNames } from './types';

/**
 * Gets the rule name to which a disable diagnostic applies. Returns `undefined`
 * if the diagnostic is not a disable diagnostic.
 * @param diagnostic The diagnostic corresponding to the Stylelint warning.
 */
export function getDisableDiagnosticRule(diagnostic: LSP.Diagnostic): string | undefined {
	// eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
	switch (diagnostic.code) {
		case DisableReportRuleNames.Needless:
			return diagnostic.message.match(/^Needless disable for "(.+)"$/)?.[1];

		case DisableReportRuleNames.InvalidScope:
			return diagnostic.message.match(/^Rule "(.+)" isn't enabled$/)?.[1];

		case DisableReportRuleNames.Descriptionless:
			return diagnostic.message.match(/^Disable for "(.+)" is missing a description$/)?.[1];

		case DisableReportRuleNames.Illegal:
			return diagnostic.message.match(/^Rule "(.+)" may not be disabled$/)?.[1];

		default:
			return undefined;
	}
}
