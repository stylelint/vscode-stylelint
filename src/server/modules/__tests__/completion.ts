import { Position, Range } from 'vscode-languageserver-types';
import { TextDocument } from 'vscode-languageserver-textdocument';
import type LSP from 'vscode-languageserver-protocol';
import { DisableReportRuleNames } from '../../../utils/stylelint/index';

import { CompletionModule } from '../completion';
import { WorkDoneProgressReporter } from 'vscode-languageserver';

const mockContext = serverMocks.getContext();
const mockLogger = serverMocks.getLogger();

const createDocument = (code: string) =>
	TextDocument.create('file:///path/test.css', 'css', 0, code);

const createNeedlessDisableDiagnostic = ({
	rule,
	range,
}: {
	rule: string;
	range: LSP.Range;
}): LSP.Diagnostic => ({
	message: `Needless disable for "${rule}"`,
	range,
	code: DisableReportRuleNames.Needless,
});

describe('CompletionModule', () => {
	beforeEach(() => {
		mockContext.__options.validate = [];
		mockContext.__options.snippet = [];
		jest.clearAllMocks();
	});

	test('should be constructable', () => {
		expect(() => new CompletionModule({ context: mockContext.__typed() })).not.toThrow();
	});

	test('onInitialize should return results', () => {
		const module = new CompletionModule({ context: mockContext.__typed() });

		expect(module.onInitialize()).toMatchSnapshot();
	});

	test('onDidRegisterHandlers should register a completion handler', () => {
		const module = new CompletionModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		expect(mockContext.connection.onCompletion).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.onCompletion).toHaveBeenCalledWith(expect.any(Function));
	});

	test('if no matching document exists, should not return completions', async () => {
		mockContext.documents.get.mockReturnValue(undefined);

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				position: Position.create(0, 0),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual([]);
		expect(mockContext.getModule).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unknown document, ignoring', { uri: 'foo' });
	});

	test('if document language ID is not in validate options, should not return completions', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['baz'];
		mockContext.__options.snippet = ['bar'];

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				position: Position.create(0, 0),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual([]);
		expect(mockContext.getModule).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Snippets or validation not enabled for language, ignoring',
			{
				uri: 'foo',
				language: 'bar',
			},
		);
	});

	test('if document language ID is not in snippet options, should not return completions', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['bar'];
		mockContext.__options.snippet = ['baz'];

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				position: Position.create(0, 0),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual([]);
		expect(mockContext.getModule).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Snippets or validation not enabled for language, ignoring',
			{
				uri: 'foo',
				language: 'bar',
			},
		);
	});

	test('with no debug log level and no valid document, should not attempt to log reason', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['bar'];
		mockContext.__options.snippet = ['baz'];
		mockLogger.isDebugEnabled.mockReturnValue(false);

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				position: Position.create(0, 0),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual([]);
		expect(mockContext.getModule).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Received onCompletion', {
			uri: 'foo',
			position: Position.create(0, 0),
		});
	});

	test('without the validator module, should return generic completions', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['bar'];
		mockContext.__options.snippet = ['bar'];
		mockContext.getModule.mockReturnValue(undefined);

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				position: Position.create(0, 0),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('without diagnostics, should return generic completions', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['bar'];
		mockContext.__options.snippet = ['bar'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				position: Position.create(0, 0),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('with no diagnostics at the same or next line, should return generic completions', async () => {
		mockContext.documents.get.mockReturnValue(createDocument('a {\n  color: red;\n}'));
		mockContext.__options.validate = ['css'];
		mockContext.__options.snippet = ['css'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [
				{
					code: 'indentation',
					message: 'Expected indentation of 4 spaces',
					range: Range.create(3, 4, 3, 4),
				},
				{
					code: 5,
					message: 'Not a Stylelint diagnostic',
					range: Range.create(1, 2, 1, 4),
				},
			],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///path/test.css' },
				position: Position.create(1, 3),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('with diagnostics at the same line, should return disable comment completions for rule', async () => {
		mockContext.documents.get.mockReturnValue(createDocument('a {\n  color: red;\n}'));
		mockContext.__options.validate = ['css'];
		mockContext.__options.snippet = ['css'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [
				{
					code: 'indentation',
					message: 'Expected indentation of 4 spaces',
					range: Range.create(1, 2, 3, 4),
				},
			],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///path/test.css' },
				position: Position.create(1, 3),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('with diagnostics at the next line, should return disable comment completions for rule', async () => {
		mockContext.documents.get.mockReturnValue(
			createDocument('a {\n  font-weight: 400;\n  color: red;\n}'),
		);
		mockContext.__options.validate = ['css'];
		mockContext.__options.snippet = ['css'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [
				{
					code: 'indentation',
					message: 'Expected indentation of 4 spaces',
					range: Range.create(2, 1, 3, 4),
				},
			],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///path/test.css' },
				position: Position.create(1, 3),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('with needless disables reported for a diagnostic, should return generic completions', async () => {
		mockContext.documents.get.mockReturnValue(createDocument('a {\n  color: red;\n}'));
		mockContext.__options.validate = ['css'];
		mockContext.__options.snippet = ['css'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [
				{
					code: 'indentation',
					message: 'Expected indentation of 4 spaces',
					range: Range.create(1, 2, 3, 4),
				},
				createNeedlessDisableDiagnostic({
					rule: 'indentation',
					range: Range.create(1, 2, 3, 4),
				}),
			],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///path/test.css' },
				position: Position.create(1, 3),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('with diagnostics at the same line and cursor in a line disable comment, should return rule completions', async () => {
		mockContext.documents.get.mockReturnValue(
			createDocument(`a {
  /* stylelint-disable-line  */
  color: red;
}`),
		);
		mockContext.__options.validate = ['css'];
		mockContext.__options.snippet = ['css'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [
				{
					code: 'indentation',
					message: 'Expected indentation of 4 spaces',
					range: Range.create(1, 2, 3, 4),
				},
			],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///path/test.css' },
				position: Position.create(1, 28),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('with diagnostics at the next line and cursor in a next-line disable comment, should return rule completions', async () => {
		mockContext.documents.get.mockReturnValue(
			createDocument(`a {
  /* stylelint-disable-next-line  */
  color: red;
}`),
		);
		mockContext.__options.validate = ['css'];
		mockContext.__options.snippet = ['css'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [
				{
					code: 'indentation',
					message: 'Expected indentation of 4 spaces',
					range: Range.create(2, 2, 3, 4),
				},
			],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///path/test.css' },
				position: Position.create(1, 33),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('with diagnostics at the next line and cursor in a disable comment, should return rule completions', async () => {
		mockContext.documents.get.mockReturnValue(
			createDocument(`a {
  /* stylelint-disable  */
  color: red;
}`),
		);
		mockContext.__options.validate = ['css'];
		mockContext.__options.snippet = ['css'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [
				{
					code: 'indentation',
					message: 'Expected indentation of 4 spaces',
					range: Range.create(2, 2, 3, 4),
				},
			],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///path/test.css' },
				position: Position.create(1, 23),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	test('with no diagnostics at same or next line and cursor in a disable comment, should not return completions', async () => {
		mockContext.documents.get.mockReturnValue(
			createDocument(`a {
    /* stylelint-disable  */
    color: red;
  font-weight: 400;
}`),
		);
		mockContext.__options.validate = ['css'];
		mockContext.__options.snippet = ['css'];
		mockContext.getModule.mockReturnValue({
			getDiagnostics: () => [
				{
					code: 'indentation',
					message: 'Expected indentation of 4 spaces',
					range: Range.create(3, 2, 3, 4),
				},
			],
			dispose: () => undefined,
		});

		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCompletion.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///path/test.css' },
				position: Position.create(1, 25),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.getModule).toHaveBeenCalledTimes(1);
		expect(result).toMatchSnapshot();
	});

	it('should be disposable', () => {
		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		expect(module).toHaveProperty('dispose');
		expect(module.dispose).toBeInstanceOf(Function);
	});

	it('should set a no-op completion handler when disposed', async () => {
		const module = new CompletionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();
		module.dispose();

		const handler = mockContext.connection.onCompletion.mock.calls[1][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				position: Position.create(0, 0),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.connection.onCompletion).toHaveBeenCalledTimes(2);
		expect(result).toBeUndefined();
	});
});
