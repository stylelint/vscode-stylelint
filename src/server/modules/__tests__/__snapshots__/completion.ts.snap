// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompletionModule onInitialize should return results 1`] = `
{
  "capabilities": {
    "completionProvider": {},
  },
}
`;

exports[`CompletionModule with diagnostics at the next line and cursor in a disable comment, should return rule completions 1`] = `
[
  {
    "detail": "disable indentation rule. (Stylelint)",
    "kind": 15,
    "label": "indentation",
  },
]
`;

exports[`CompletionModule with diagnostics at the next line and cursor in a next-line disable comment, should return rule completions 1`] = `
[
  {
    "detail": "disable indentation rule. (Stylelint)",
    "kind": 15,
    "label": "indentation",
  },
]
`;

exports[`CompletionModule with diagnostics at the next line, should return disable comment completions for rule 1`] = `
[
  {
    "detail": "Turn off Stylelint rules for individual lines only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-line",
  },
  {
    "detail": "Turn off Stylelint rules for the next line only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-next-line indentation */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-next-line \${0:indentation} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-next-line",
  },
  {
    "detail": "Turn off all Stylelint or individual rules, after which you do not need to re-enable Stylelint. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable rule */
/* stylelint-enable rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable \${0:rule} */
/* stylelint-enable \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable",
  },
]
`;

exports[`CompletionModule with diagnostics at the same line and cursor in a line disable comment, should return rule completions 1`] = `
[
  {
    "detail": "disable indentation rule. (Stylelint)",
    "kind": 15,
    "label": "indentation",
  },
]
`;

exports[`CompletionModule with diagnostics at the same line, should return disable comment completions for rule 1`] = `
[
  {
    "detail": "Turn off Stylelint rules for individual lines only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-line indentation */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-line \${0:indentation} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-line",
  },
  {
    "detail": "Turn off Stylelint rules for the next line only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-next-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-next-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-next-line",
  },
  {
    "detail": "Turn off all Stylelint or individual rules, after which you do not need to re-enable Stylelint. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable rule */
/* stylelint-enable rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable \${0:rule} */
/* stylelint-enable \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable",
  },
]
`;

exports[`CompletionModule with needless disables reported for a diagnostic, should return generic completions 1`] = `
[
  {
    "detail": "Turn off Stylelint rules for individual lines only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-line",
  },
  {
    "detail": "Turn off Stylelint rules for the next line only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-next-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-next-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-next-line",
  },
  {
    "detail": "Turn off all Stylelint or individual rules, after which you do not need to re-enable Stylelint. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable rule */
/* stylelint-enable rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable \${0:rule} */
/* stylelint-enable \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable",
  },
]
`;

exports[`CompletionModule with no diagnostics at same or next line and cursor in a disable comment, should not return completions 1`] = `[]`;

exports[`CompletionModule with no diagnostics at the same or next line, should return generic completions 1`] = `
[
  {
    "detail": "Turn off Stylelint rules for individual lines only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-line",
  },
  {
    "detail": "Turn off Stylelint rules for the next line only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-next-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-next-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-next-line",
  },
  {
    "detail": "Turn off all Stylelint or individual rules, after which you do not need to re-enable Stylelint. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable rule */
/* stylelint-enable rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable \${0:rule} */
/* stylelint-enable \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable",
  },
]
`;

exports[`CompletionModule without diagnostics, should return generic completions 1`] = `
[
  {
    "detail": "Turn off Stylelint rules for individual lines only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-line",
  },
  {
    "detail": "Turn off Stylelint rules for the next line only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-next-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-next-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-next-line",
  },
  {
    "detail": "Turn off all Stylelint or individual rules, after which you do not need to re-enable Stylelint. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable rule */
/* stylelint-enable rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable \${0:rule} */
/* stylelint-enable \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable",
  },
]
`;

exports[`CompletionModule without the validator module, should return generic completions 1`] = `
[
  {
    "detail": "Turn off Stylelint rules for individual lines only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-line",
  },
  {
    "detail": "Turn off Stylelint rules for the next line only, after which you do not need to explicitly re-enable them. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable-next-line rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable-next-line \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable-next-line",
  },
  {
    "detail": "Turn off all Stylelint or individual rules, after which you do not need to re-enable Stylelint. (Stylelint)",
    "documentation": {
      "kind": "markdown",
      "value": "\`\`\`css
/* stylelint-disable rule */
/* stylelint-enable rule */
\`\`\`",
    },
    "insertText": "/* stylelint-disable \${0:rule} */
/* stylelint-enable \${0:rule} */",
    "insertTextFormat": 2,
    "kind": 15,
    "label": "stylelint-disable",
  },
]
`;
